File: scripts/julia_controller/macros.jl
# macros.jl - Record/replay motor-pose macros, saved as JSON. Julia
# counterpart of controller.py's macro helpers - reads/writes the same
# scripts/macros/*.json files as the Python and C++ controllers.

import JSON3
using Dates

const MACROS_DIR = normpath(joinpath(@__DIR__, "..", "macros"))

struct MacroStep
    t::Float64
    commands::Dict{Int,Int}   # channel -> angle
end

function list_macros()
    isdir(MACROS_DIR) || return String[]
    files = [f for f in readdir(MACROS_DIR; join=true) if endswith(f, ".json")]
    return sort(files; rev=true)
end

function save_macro(steps::Vector{MacroStep})
    mkpath(MACROS_DIR)
    path = joinpath(MACROS_DIR, "macro_$(Dates.format(now(), "yyyymmdd_HHMMSS")).json")
    payload = Dict(
        "created" => Dates.format(now(), "yyyy-mm-dd HH:MM:SS"),
        "steps" => [
            Dict("t" => s.t, "commands" => Dict(string(ch) => ang for (ch, ang) in s.commands))
            for s in steps
        ],
    )
    open(path, "w") do io
        JSON3.write(io, payload)
    end
    return path
end

function load_macro(path)
    data = JSON3.read(read(path, String), Dict{String,Any})
    steps = MacroStep[]
    for step in get(data, "steps", [])
        t = Float64(step["t"])
        commands = Dict{Int,Int}(parse(Int, ch) => round(Int, ang) for (ch, ang) in step["commands"])
        push!(steps, MacroStep(t, commands))
    end
    return steps
end

File: scripts/go_controller/macros.go
package main

// Record/replay motor-pose macros as JSON. Go counterpart of controller.py's
// macro helpers; reads/writes the same scripts/macros/*.json files as every
// other controller.

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"time"
)

type MacroStep struct {
	T        float64
	Commands map[int]int // channel -> angle
}

// listMacros returns macro files newest first (file names embed the timestamp).
func listMacros() []string {
	files, _ := filepath.Glob(filepath.Join(macrosDir(), "*.json"))
	sort.Slice(files, func(i, j int) bool { return filepath.Base(files[i]) > filepath.Base(files[j]) })
	return files
}

func saveMacro(steps []MacroStep) (string, error) {
	if err := os.MkdirAll(macrosDir(), 0o755); err != nil {
		return "", err
	}
	now := time.Now()
	path := filepath.Join(macrosDir(), "macro_"+now.Format("20060102_150405")+".json")

	stepList := make([]map[string]any, 0, len(steps))
	for _, s := range steps {
		commands := map[string]int{}
		for ch, a := range s.Commands {
			commands[strconv.Itoa(ch)] = a
		}
		stepList = append(stepList, map[string]any{"t": s.T, "commands": commands})
	}
	data, err := json.Marshal(map[string]any{"created": now.Format("2006-01-02 15:04:05"), "steps": stepList})
	if err != nil {
		return "", err
	}
	return path, os.WriteFile(path, data, 0o644)
}

func loadMacro(path string) ([]MacroStep, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var root struct {
		Steps []struct {
			T        float64            `json:"t"`
			Commands map[string]float64 `json:"commands"`
		} `json:"steps"`
	}
	if err := json.Unmarshal(data, &root); err != nil {
		return nil, err
	}
	steps := make([]MacroStep, 0, len(root.Steps))
	for _, s := range root.Steps {
		commands := map[int]int{}
		for key, a := range s.Commands {
			if ch, err := strconv.Atoi(key); err == nil {
				commands[ch] = int(math.Round(a))
			}
		}
		steps = append(steps, MacroStep{T: s.T, Commands: commands})
	}
	return steps, nil
}
